core,rp2,esp8266,windows, unix: Add new cross-port functions for event waiting and handling.

projectgus opened the pull request:
This PR fixes #12925, which is a regression on rp2 introduced when rp2 became "tickless" in #12837.

It implements one of the [suggestions from the linked Issue](https://github.com/micropython/micropython/issues/12925#issuecomment-1803038430): to add new functions for "wait for event" and "handle events". The new functions are port-agnostic:

* `mp_event_handle_nowait()` - Handle any pending MicroPython events without waiting for an interrupt or event.
* `mp_event_wait_indefinite()` - Handle any pending MicroPython events and then suspend execution until the next interrupt or event. This function is deliberately named to attract attention when developers use it, as on tickless ports it will cause issues if the developer expects it to return without a corresponding interrupt having occurred.
* `mp_event_wait_ms(mp_uint_t timeout_ms)` - Handle any pending MicroPython events and then suspend execution until the next interrupt or event, or until timeout_ms milliseconds have elapsed.

Each port only needs to provide a `MICROPY_INTERNAL_WFE(OPTIONAL_TIMEOUT)` macro to implement the "wait for event" logic. More port-specific macros may be needed when adding other ports.

These functions are intended to eventually replace `MICROPY_EVENT_POLL_HOOK` and `MICROPY_EVENT_POLL_HOOK_FAST`, although on all ports except rp2 they are currently implemented using these macros.

This PR switches over `extmod`, `rp2` port and the built-in drivers supported by `rp2` to use the new functions, fixing the issue of indefinite sleep under some conditions.

It also switches over `esp8266`, to avoid problems where `i2c.scan()` doesn't call `ets_event_poll()` and can trigger a WDT.

Finally it switches over Windows, as this was necessary to get the same sleep behaviour with the new functions, and unix  so that the coverage tests pass on the new functions.

`rp2` port also enables [SEVONPEND](https://developer.arm.com/documentation/dui0497/a/cortex-m0-peripherals/system-control-block/system-control-register) to ensure `SEV` bit is set on each interrupt, removing potential race conditions between the interrupt and the main thread checking a hardware condition. It may be possible to make this more fine-grained in future by manually calling `SEV` from individual ISRs, but this approach ensures the firmware can't become stuck if a `SEV` is missing.

*This work was funded through GitHub Sponsors.*

projectgus committed:
extmod: Switch to use new event functions.

See previous commit for details of these functions.  As of this commit,
these still call the old hook macros on all ports.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
rp2: Switch rp2 and drivers to use new event functions.

This commit changes all uses in the rp2 port, and drivers that are
optionally supported by that port.

The old MICROPY_EVENT_POLL_HOOK and MICROPY_EVENT_POLL_HOOK_FAST macros are
no longer used for rp2 builds and are removed (C user code will need to be
changed to suit).

Also take the opportunity to change some timeouts that used 64-bit
arithmetic to 32-bit, to hopefully claw back a little code size.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
py: Add port-agnostic inline functions for event handling.

These are intended to replace MICROPY_EVENT_POLL_HOOK and
MICROPY_EVENT_POLL_HOOK_FAST, which are insufficient for tickless ports.

This implementation is along the lines suggested here:
https://github.com/micropython/micropython/issues/12925#issuecomment-1803038430

Currently any usage of these functions expands to use the existing hook
macros, but this can be switched over port by port.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
rp2/main: Enable SEVONPEND CPU interrupt bit.

Previously this was not set, so potential for race conditions in interrupt
handlers this didn't issue SEV.  (Which is currently all of them, as far as
I can see.)

Eventually we might be able to augment the interrupt handlers that wake the
main thread to call SEV, and leave the others as-is to suspend the CPU
slightly faster, but this will solve the issue for now.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
windows: Implement MICROPY_INTERNAL_WFE() macro.

This should be the equivalent of the previous event poll hook macro.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Hmm, so the challenge is making these functions instead of macros is ports with threading.

A lot of the ports with threading call `MP_THREAD_GIL_EXIT()` & `MP_THREAD_GIL_ENTER()` as part of their wait for event hook. These macros are defined in `py/mpthread.h` and they depend on some other headers as well.

This creates a layering violation with `mphal.h` as it's a pretty low level header. The layering violation doesn't matter for macro expansion, provided all of the relevant parts have been included when the macro expands.

Options would seem to be:

1. Make these back into macros.
2. Put these functions somewhere else, a new header like `py/event.h` or perhaps `py/runtime.h` (which contains the closest related functions, from what I can see.)

A version of the second option seems better from the point of view of having headers that "include what they use" and layer nicely, but the first option will be quick and easy...

projectgus commented:
Oh, I thought of a third option: Convert `mp_wait_for_event_indefinite()` and `mp_wait_for_event(mp_uint_t timeout_ms)` to normal functions instead of inline. Then they can include whatever they need, and there will be a small code size win.

(Unlike the "nowait" version, these generally get called in places where I don't expect the overhead of one more function call will be very noticeable.)

dpgeorge commented on py/mphal.h:
```diff
@@ -26,7 +26,9 @@
 
+#include <stddef.h>
 #include <stdint.h>
+#include <stdbool.h>
```
Do these new includes need to be added?

projectgus replied:
No, this was a leftover. Thanks!

dpgeorge commented on py/runtime.h:
```diff
@@ -26,6 +26,7 @@
 #ifndef MICROPY_INCLUDED_PY_RUNTIME_H
 #define MICROPY_INCLUDED_PY_RUNTIME_H
 
+#include "py/mphal.h"
```
This breaks the example/natmod builds... I think just remove it from here.  This header should only be included from .c files.

projectgus replied:
This is tricky. Without a definition of `MICROPY_INTERNAL_WFE`, including `runtime.h` will fail (the difference between macros and inline functions is they force you to structure your headers with a more layered "include what you use"!)

Conceptually these "event" functions are a bit different, in as much as runtime.h is "pure Python runtime", mphal.h is "hardware abstractions", and then event functions are a bridge between the two.

I've updated this PR to put them into a new `py/event.h` header, which solves these kind of conflicts (and allows removing all the messing about with esp8266 headers, too.)

Let me know what you think.

dpgeorge replied:
I'm pretty sure I tried just removing this include and the build worked.... but maybe I didn't test enough.

How about putting these inline functions in `py/mphal.h`?  As you say, they are half hardware abstraction functions.

dpgeorge replied:
> How about putting these inline functions in `py/mphal.h`? As you say, they are half hardware abstraction functions.

Hmm, maybe that's a bad idea, because then `py/mphal.h` will need to include `py/runtime.h`, which is no good.

projectgus replied:
Yeah, that's it. My very first version of this added the inlines to`mphal.h` and ran into that problem.

I think conceptually:

* `mphal.h` is "lower" and relates to the hardware.
* `runtime.h` is "higher" and relates to the pure Python layer.
* `event.h` is "in between" the two.

dpgeorge commented on ports/rp2/cyw43_configport.h:
```diff
@@ -119,6 +119,6 @@ static inline void cyw43_delay_ms(uint32_t ms) {
 }
 
-#define CYW43_EVENT_POLL_HOOK MICROPY_EVENT_POLL_HOOK_FAST
+#define CYW43_EVENT_POLL_HOOK mp_event_handle_nowait()
```
this file now needs to include runtime.h, to get this function defn

dpgeorge commented:
I've tested this on a Pico and Pico W and they both now pass all the asyncio tests (again).  Also networking and BLE works on Pico W.

dpgeorge commented on py/runtime.h:
```diff
@@ -109,6 +110,46 @@ bool mp_sched_schedule(mp_obj_t function, mp_obj_t arg);
 
+// Handles any pending MicroPython events without waiting for an interrupt or
+// event.
+STATIC MP_ALWAYSINLINE inline void mp_event_handle_nowait(void) {
```
When using static-inline the static must always be there, so STATIC should be static.

Write this either as:
```
MP_ALWAYSINLINE static inline ...
```
or:
```
static inline MP_ALWAYSINLINE ...
```

projectgus replied:
Oh, I wondered why MicroPython had this `STATIC` macro. Is the idea that sometimes you want to build with more visible symbols?

projectgus replied:
There's even a comment that explains it! I should have looked.

projectgus replied:
Resolved (in py/event.h now)

dpgeorge replied:
> Is the idea that sometimes you want to build with more visible symbols?

Yes, exactly.  Although I suspect there are better ways to do that with compiler/linker flags.  So I'd prefer to eventually move away from using our `STATIC` macro.

dpgeorge commented on py/runtime.h:
```diff
@@ -109,6 +110,46 @@ bool mp_sched_schedule(mp_obj_t function, mp_obj_t arg);
+// next interrupt or event.
+//
+// Note: on "tickless" ports this can suspend execution for a long time,
+// don't call unless you know an interrupt is coming to continue execution.
```
Maybe add "and may also return early".

projectgus replied:
Reworded both these comments to explicitly explain that it's the tick macro that can cause frequent wakeups.

dpgeorge commented on py/runtime.h:
```diff
@@ -109,6 +110,46 @@ bool mp_sched_schedule(mp_obj_t function, mp_obj_t arg);
+// Handle any pending MicroPython events and then suspends execution until the
+// next interrupt or event, or until timeout_ms milliseconds have elapsed.
+//
+// This function will often return early.
```
Maybe phrase it as "may return early".

dpgeorge commented on ports/windows/mpconfigport.h:
```diff
@@ -229,12 +229,9 @@ typedef long mp_off_t;
-        mp_handle_pending(true); \
-        msec_sleep(1.0); \
-    } while (0);
+#define MICROPY_INTERNAL_WFE(TIMEOUT_MS) msec_sleep(MIN(1.0, (double)(TIMEOUT_MS)))
```
What happens when TIMEOUT_MS is -1?

projectgus replied:
Derp this should be MAX so it always sleeps at least 1ms, good catch.

dpgeorge commented on ports/rp2/mpconfigport.h:
```diff
@@ -249,17 +249,16 @@ extern const struct _mp_obj_type_t mod_network_nic_type_wiznet5k;
-        MICROPY_EVENT_POLL_HOOK_FAST; \
-        __wfe(); \
-    } while (0);
+// Port level Wait for Event macro
```
"Wait-For-Event"

dpgeorge commented on ports/rp2/mpconfigport.h:
```diff
@@ -249,17 +249,16 @@ extern const struct _mp_obj_type_t mod_network_nic_type_wiznet5k;
+        } else { \
+            best_effort_wfe_or_timeout(make_timeout_time_ms(TIMEOUT_MS)); \
+        } \
+} while (0)
```
please indent this last line by 4 spaces

projectgus replied:
Each time I change this, uncrustify (in pre-commit) is changing it back...

dpgeorge commented on ports/esp8266/esp_mphal.h:
```diff
@@ -24,6 +24,9 @@
  */
 
+#ifndef _ESP8266_ESP_MPHAL_H
+#define _ESP8266_ESP_MPHAL_H
```
please use `MICROPY_INCLUDED_ESP8266_ESP_MPHAL_H`

projectgus replied:
Dropped this commit entirely, but I'll submit this as it's own small PR

projectgus replied:
Over here now: https://github.com/micropython/micropython/pull/13147

dpgeorge commented:
I think esp8266 needs to be updated to use the new scheme, because it does things differently to the other ports:
```
#define MICROPY_EVENT_POLL_HOOK {ets_event_poll();}
```
In particular the `i2c.scan()` may cause a watchdog timeout if there are no pull-up resistors on the bus.

dpgeorge commented on py/runtime.h:
```diff
@@ -109,6 +110,46 @@ bool mp_sched_schedule(mp_obj_t function, mp_obj_t arg);
+    // For ports still using the old macros
+    MICROPY_EVENT_POLL_HOOK_FAST
+    #else
+    mp_handle_pending(true);
```
This is not general enough.  Some ports like esp8266 must do work in addition to `mp_handle_pending()`.  So we need a way to hook into this function with a macro.

I guess for now we could use `MICROPY_EVENT_POLL_HOOK_FAST` for this, but eventually we will need a way to hook.

projectgus replied:
> In the specific case of MICROPY_EVENT_POLL_HOOK_FAST and mp_event_handle_no_wait(), esp8266 port doesn't define this hook at the moment so I think it will keep working much the same?

Ah, I'm with you now that the i2c scan specifically failed over to the "slow" hook if the fast hook wasn't defined.

OK, will check all the ports for this behaviour.

projectgus replied:
Updated esp8266 with a new hook `MICROPY_INTERNAL_EVENT_HOOK` and verified i2c.scan() completes. Do you know if any other ports relied on that fallback logic in i2c? It looks like probably not.

Compared to before, there is a small behaviour change - esp8266 port will call `ets_event_poll()` anywhere else that `mp_event_handle_nowait()` is called. I think this is only modselect.c though, and it seems like a good idea to be processing wifi events in that instance.

dpgeorge replied:
> Ah, I'm with you now that the i2c scan specifically failed over to the "slow" hook if the fast hook wasn't defined.

Indeed.  The existing behaviour is quite subtle, actually too subtle.  I'm pretty sure only esp8266 relied on it.



> Compared to before, there is a small behaviour change - esp8266 port will call `ets_event_poll()` anywhere else that `mp_event_handle_nowait()` is called. I think this is only modselect.c though, and it seems like a good idea to be processing wifi events in that instance.

That sounds like a good improvement.

projectgus committed:
esp8266: Update port to use new event functions.

This is necessary to avoid watchdog timeout in long i2c.scan(), as
previously machine_i2c.c would call MICROPY_EVENT_POLL_HOOK if
MICROPY_EVENT_POLL_HOOK_FAST was not available.

Compared to previous implementation, this implementation removes the
ets_event_poll() function and calls the SDK function ets_loop_iter() from
MICROPY_INTERNAL_EVENT_HOOK instead.  This allows using the port-agnostic
functions in more places.

There is a small behaviour change, which is that the event loop gets
iterated in a few more places (i.e. anywhere that mp_event_handle_nowait()
is called).  However, this looks like maybe only modselect.c - and is
probably good to process Wi-Fi events in that polling loop.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

dpgeorge commented:
There's a problem with the AppVeyor CI:
```
c:\projects\micropython\py\event.h(34): warning C4141: 'inline': used more than once (compiling source file C:\projects\micropython\extmod\modselect.c) [C:\projects\micropython\ports\windows\micropython.vcxproj]
         c:\projects\micropython\py\event.h(54): warning C4141: 'inline': used more than once (compiling source file C:\projects\micropython\extmod\modselect.c) [C:\projects\micropython\ports\windows\micropython.vcxproj]
         c:\projects\micropython\py\event.h(68): warning C4141: 'inline': used more than once (compiling source file C:\projects\micropython\extmod\modselect.c) [C:\projects\micropython\ports\windows\micropython.vcxproj]
```

dpgeorge commented on ports/esp8266/machine_uart.c:
```diff
@@ -300,7 +301,7 @@ STATIC mp_uint_t mp_machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, uint
                 return 0;
             }
-            MICROPY_EVENT_POLL_HOOK
+            mp_event_wait_indefinite();
```
This should be `mp_event_wait_ms(1)`, since the loop checks for a timeout.

projectgus replied:
Oops, good catch! (I think they're equivalent on this port, but good to use the correct function anyhow.)

dpgeorge commented on ports/esp8266/mpconfigport.h:
```diff
@@ -116,7 +116,11 @@
+// No blocking wait-for-event on ESP8266, only non-blocking poll of the OS event
+// loop
+#define MICROPY_INTERNAL_WFE(TIMEOUT_MS)
+#define MICROPY_INTERNAL_EVENT_HOOK ets_event_poll()
```
The `ets_event_poll()` function is implemented as:
```c
void ets_event_poll(void) {
    ets_loop_iter();
    mp_handle_pending(true);
}
```
It already calls `mp_handle_pending(true)`, so now will call it twice in `mp_event_xxx()` functions.

I suggest changing this macro to call `ets_loop_iter()` directly, remove `ets_event_poll()`, and for the small number of uses of `ets_event_poll()` change them to `mp_event_handle_nowait()`.

projectgus commented on py/event.h:
```diff
@@ -0,0 +1,78 @@
+
+// Handles any pending MicroPython events without waiting for an interrupt or
+// event
+static inline MP_ALWAYSINLINE void mp_event_handle_nowait(void) {
```
Good point, the compiler can be a pretty good judge of these.

(The `mp_event_wait_` ones could even arguably be made into regular functions I think.)

dpgeorge replied:
I've maid them all into functions.

projectgus committed:
unix: Update port to use the new event functions.

Necessary to get coverage of the new event functions.

Deletes the case that called usleep(delay) for mp_hal_delay_ms(), it seems
like this wouldn't have ever happened anyhow (MICROPY_EVENT_POOL_HOOK is
always defined for the unix port).

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge Coverage is still failing. I think because the only code which calls `mp_event_wait_ms()` on the unix port is `mp_hal_delay_ms`. On unix+float, `time.sleep()` is implemented by calling `sleep_select()`.

It looks like `select()` has been used for sleep with floating point arguments for a long time, since f7662645821d4528f0e36a51ed8d41749748230e and refined in 0bb57bf5bf1c9822abca524aabd00fe9b861d10d.

It seems like it might be reasonable to refactor this to call `mp_hal_delay_ms()` in all cases, though? It'll use a modicum more CPU time, but not a lot.

(The other option might be to implement `MICROPY_INTERNAL_WFE` on unix by calling select, as EINTR causes select to return early, but that seems a lot fiddlier.)

dpgeorge commented:
> I think because the only code which calls `mp_event_wait_ms()` on the unix port is `mp_hal_delay_ms`. On unix+float, `time.sleep()` is implemented by calling `sleep_select()` not `mp_hal_delay_ms()`.

OK, that's a bit tricky.

How about this then:
```diff
--- a/ports/unix/coverage.c
+++ b/ports/unix/coverage.c
@@ -578,9 +578,9 @@ STATIC mp_obj_t extra_coverage(void) {
         mp_sched_unlock();
         mp_printf(&mp_plat_print, "unlocked\n");
 
-        // drain pending callbacks
+        // drain pending callbacks (and test mp_event_wait_ms())
         while (mp_sched_num_pending()) {
-            mp_handle_pending(true);
+            mp_event_wait_ms(1);
         }
 
         // setting the keyboard interrupt and raising it during mp_handle_pending
```

dpgeorge commented:
> Oh, I thought of a third option: Convert `mp_wait_for_event_indefinite()` and `mp_wait_for_event(mp_uint_t timeout_ms)` to normal functions instead of inline. Then they can include whatever they need, and there will be a small code size win.

I've changed it so all three of these event functions are normal functions (not inline).

This saves code size, and makes the abstraction a bit stronger.  These functions really don't need to be inline.  The original macros were because that was a bit of a hack that just allowed a port to insert arbitrary code for the polling.  But now that it has been formalised in this PR these event functions are better not inlined.  If they are inline then the more they are used the more code size will grow, and we want to be able to use them on all ports, even ones that are space constrained.

I checked all uses of these event functions and none of them are critical that they warrant (or need, eg to be in IRAM) an inline function.  Better to spend the bytes elsewhere to improve performance.  It also means we don't need a new `py/event.h` header file.

stinos commented on ports/windows/mpconfigport.h:
```diff
@@ -227,14 +227,11 @@ typedef long mp_off_t;
-    } while (0);
+#define MICROPY_INTERNAL_WFE(TIMEOUT_MS) msec_sleep(MAX(1.0, (double)(TIMEOUT_MS)))
+#else
+#define MICROPY_INTERNAL_WFE(TIMEOUT_MS) /* No-op */
```
Not completely sure what MICROPY_INTERNAL_WFE is supposed to do, but scheduler.c has `MICROPY_INTERNAL_WFE(-1)` so that does not seem to work as intended?

dpgeorge replied:
MICROPY_INTERNAL_WFE is "wait for event".  But it can be conservative and return immediately, but in that case it'll use more power/CPU.

If the argument is -1 it means "wait indefinitely for the next event".  Here that would just sleep 1ms.

When MICROPY_ENABLE_SCHEDULER is disabled, I don't think this macro is ever used, so doesn't need to do anything in that case.
